### PR TITLE
KAFKA-13922: Adjustments for jacoco, coverage reporting

### PR DESCRIPTION
**Pull request by akatona84**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -998,6 +973,9 @@ project(':core') {
   if (userEnableTestCoverage) {
     scoverage {
       scoverageVersion = versions.scoverage
+      if (versions.baseScala == '2.13') {
+        scoverageScalaVersion = '2.13.8' // there's no 2.13.10 artifact, org.scoverage:scalac-scoverage-plugin_2.13.8:1.4.11 is the latest as of now
```
Comment (viktorsomogyi) [inline build.gradle]: Does this mean that the coverage will use 2.13.8 while compilation still uses 2.13.10 or whatever is set by gradle?

File: build.gradle [hunk 2]
```diff
@@ -1289,7 +1267,7 @@ project(':metadata') {
     }
     test {
       java {
-        srcDirs = ["src/generated/java", "src/test/java"]
+        srcDirs = ["src/test/java"]
```
Comment (viktorsomogyi) [inline build.gradle:1275]: Why did you need to remove the generated sources? Can't we just exclude these packages from the coverage report?

File: gradle/dependencies.gradle [hunk 3]
```diff
@@ -158,7 +158,7 @@ versions += [
   // https://github.com/scalameta/scalafmt/releases/tag/v3.1.0.
   scalafmt: "3.7.14",
   scalaJava8Compat : "1.0.2",
-  scoverage: "1.9.3",
+  scoverage: "1.4.11",
```
Comment (viktorsomogyi) [inline gradle/dependencies.gradle]: Why did you need to downgrade this version?

File: build.gradle [hunk 4]
```diff
@@ -799,33 +799,8 @@ def checkstyleConfigProperties(configFileName) {
   [importControlFile: "$configFileName"]
 }
 
-// Aggregates all jacoco results into the root project directory
```
Comment (viktorsomogyi) [inline build.gradle:803]: Where would one use this functionality and why don't we need it?

File: gradle/dependencies.gradle (hunk 3 above)
Comment (akatona84) [reply gradle/dependencies.gradle]: this is the last 1.x version with 2.13 scala
https://repo1.maven.org/maven2/org/scoverage/scalac-scoverage-plugin_2.13.8/

there's one more for 2.13 scala: 2.13.9, but it only has 2.x versions under that
https://repo1.maven.org/maven2/org/scoverage/scalac-scoverage-plugin_2.13.9/

File: build.gradle (hunk 2 above)
Comment (akatona84) [reply build.gradle:1275]: it's not exclusion, it's deduplication 
so it's not strictly conflicts with coverage, but since the src/generated/java is already there in sourceSets.main.java it is unnecessary to repeat it in test.java sourceset as well.

Sonar plugin was the one which wasn't working until these were cleared out.

File: build.gradle (hunk 4 above)
Comment (akatona84) [reply build.gradle:803]: This part is not working for a while, gradle+jacoco upgrades made it incompatible with them.
When it worked, it was created an aggregated html coverage report, after the tests with coverage ran on the  whole project. somewhat useful for the human eye to see all coverage at one place. Jenkins doesn't care about this report and it aggregates the submodules' reports, sonar: don't care either. Individual reports can be viewed under each modules report dir.

I could not fix this aggregated root report or replace it, spent quite a lot of time on it but felt it doesn't worth it. Since this root report doesn't work at all, better have individual coverage reports per module than none.

File: build.gradle (hunk 1 above)
Comment (akatona84) [reply build.gradle]: I'm not sure how it works actually, but assuming it's pulling in another dependency where the transitive scala deps are overridden by the newer, currently set one. Or it's restricted only for the plugin.
will check

Comment (akatona84) [reply build.gradle]:
```
scoverage - Scoverage dependencies
+--- org.scoverage:scalac-scoverage-plugin_2.13.8:1.4.11
|    +--- org.scala-lang:scala-library:2.13.8 -> 2.13.12
|    \--- org.scala-lang.modules:scala-xml_2.13:1.3.0
|         \--- org.scala-lang:scala-library:2.13.1 -> 2.13.12
\--- org.scoverage:scalac-scoverage-runtime_2.13:1.4.11
     \--- org.scala-lang:scala-library:2.13.8 -> 2.13.12
```

File: gradle/dependencies.gradle (hunk 3 above)
Comment (akatona84) [reply gradle/dependencies.gradle]: I'm trying to upgrade to 2.x scoverage (with plugin upgrade too)

File: build.gradle (hunk 4 above)
Comment (viktorsomogyi) [reply build.gradle:803]: Ok, that sounds good. I think the next step to this PR would be to actually enable the jacoco plugin in jenkins, I'll look into how can we do that.

**Review by viktorsomogyi — APPROVED** [review]
Overall I think these changes are good. The test failures don't matter as the change affects only report generation.
It would be nice to get an aggregated report to keep the existing functionality (and for the eye), but since neither Jenkins nor Sonar cares about it, I think it can be done in an optional future PR.

I'll merge this soon.

One more thing: it'd be nice to enable the coverage report in Jenkins but for that we'd need to change a few things in Jenkinsfile, disable parallel builds and use the `reportCoverage` target instead of `test` and `integrationTest`, which I think should be discussed with the wider community as it affects the speed of builds.
